### PR TITLE
Various changes to the reconnect strategy of TCP provider

### LINKED ISSUE #86: TCP provider doesn't always reconnect
(1) The TCP provider doesn't reconnect unless reconnect is explicitly set to `true` in config, due to a OR issue. 

(2) If the TCP server disconnects, `stream.prototype.end` is called on TCP provider. This sets `reconnect` to false, causing the provider to never reconnect with the TCP server.

### PR DESCRIPTION
**Pull request by fabdrol**
This PR fixes #86

### CODE REVIEW

**Comment by fabdrol** [comment]
@tkurki any objections? Otherwise I'd like to merge :)

File: providers/tcp.js [hunk 1]
```diff
@@ -20,46 +20,47 @@ var net       = require('net')
...
+    debug('Stream: "error". Message: ' + err.message);
   });
 }
 
 require('util').inherits(TcpStream, Transform);
 
-TcpStream.prototype.handleTimeout = function() {
-  if((Date.now() - this.__last) > 90000) {
+TcpStream.prototype.handleTimeout = function () {
+  if ((Date.now() - this.__last) > 90000 && this.__reset === null) {
```
Comment (tkurki) [inline providers/tcp.js:49]: Won't this cause reconnects when tcp connection is ok but there is no data coming over the connection?

File: providers/tcp.js [hunk 2]
```diff
@@ -68,47 +69,65 @@ TcpStream.prototype.start = function() {
     this.socket = null;
   }
 
-  if(this.reconnect === true && this.socket === null) {
-    if(this.__timeout !== null) {
-      clearTimeout(this.__timeout);
-    }
+  if (this.socket !== null) {
+    debug('Fatal: socket cannot be null at this point. Game over.')
```
Comment (tkurki) [inline providers/tcp.js]: Must be null?

File: providers/tcp.js [hunk 3]
```diff
@@ -68,47 +69,65 @@ TcpStream.prototype.start = function() {
...
+    return
+  }
+
+  if (this.reconnect !== true) {
```
Comment (tkurki) [inline providers/tcp.js]: If reconnect is false start() will never do anything.

**Comment by tkurki** [comment]
Wouldn't it be easier to use something like https://www.npmjs.com/package/net-socket-reconnect (just googled, no experience with this or other npm solutions).

File: providers/tcp.js (hunk 1 above)
Comment (fabdrol) [inline providers/tcp.js:49]: Specifically, when the `data` event is not fired for 90 seconds. It's intended to catch certain eventualities where the TCP stream seems open, but no data is received, when there should be. Frankly I didn't think about that situation the other way round (perfectly healthy stream, but no data).

File: providers/tcp.js (hunk 3 above)
Comment (fabdrol) [inline providers/tcp.js]: that's an error - will change.

**Comment by tkurki** [comment]
If you're happy with this go ahead and merge.

**Pull request merged**